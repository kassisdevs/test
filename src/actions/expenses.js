import uuid from 'uuid';

// ADD_EXPENCE action
export const ADD_EXPENSE = ({
  id = uuid(),
  description = '',
  note = '',
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
  id,
  description,
  note,
  amount,
  createdAt
  }
});

// EDIT EXPENCES
export const EDIT_EXPENSE = ({
  id,
  updates
}) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})
// REMOVE_EXPENSES
export const REMOVE_EXPENSE = (
{
id
} = {}
) => ({
  type: 'REMOVE_EXPENSE',
  id
})