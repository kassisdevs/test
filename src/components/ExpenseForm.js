import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarfocused: false,
      error: ''
    };
  }
  onChangeDes = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onChangeNote = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }));
  };
  onChangeAmount = (e) => {
    const amount = e.target.value
  if (amount.match(/^\d*(\.\d{0,2})?$/)) {
    this.setState(() => ({ amount }))
    }
  };
  onChangeDate = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarfocused: focused }));
  };
  onSubmitForm = (e) => {
    if (!this.state.description || !this.state.amount) {
     this.setState(() => ({ error: 'Please enter a valid description and amount' }));
    } else {
     this.setState(() => ({ error: '' }));
     this.props.onSubmit({
      description: this.state.description,
      amount: Number(this.state.amount) * 100,
      note: this.state.note,
      createdAt: this.state.createdAt.valueOf()
     });
    }
    e.preventDefault();
  }
  render() {
    return (
      <div>
      {this.state.error && <p>{this.state.error}</p>}
      <form onSubmit={this.onSubmitForm}>
      <input
      type="text"
      value={this.state.description}
      placeholder="description"
      onChange={this.onChangeDes}
      />
      <input
      type="number"
      placeholder="amount"
      value={this.state.amount}
      onChange={this.onChangeAmount}
      />
      <SingleDatePicker
       date={this.state.createdAt}
       onDateChange={this.onChangeDate}
       focused={this.state.calendarfocused}
       onFocusChange={this.onFocusChange}
       numberOfMonths={1}
       isOutsideRange={() => false}
      />
      <textarea
      placeholder="note"
      value={this.state.note}
      onChange={this.onChangeNote}
      >
      </textarea>
      <button
      
      >
      Add Expense</button>

      </form>
      </div>
    )
  }
}