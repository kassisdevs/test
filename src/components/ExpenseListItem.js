import React from 'react';
import { Link } from 'react-router-dom'


export const ExpenseListItem = ({ dispatch, id, description, note, amount, createdAt }) => (
<div>
<Link to={`/edit/${id}`}>
<h1>{description}</h1>
</Link>
<p>{note}</p>
<p>{amount} - {createdAt}</p>
</div>
);


export default ExpenseListItem;