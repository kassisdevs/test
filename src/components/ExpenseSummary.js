import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectTotalExpenses from '../selectors/expenses-total';

export const ExpenseSammary = (props) => (
  <div>
  <p>
  Viewing {props.expenseCount} expenses totaling {numeral(props.expensesTotal / 100).format('$0,0.00')}
  </p>
  </div>
);

const mapToStateStore = (state) => ({
  expenseCount: state.expenses.length,
  expensesTotal: selectTotalExpenses(state.expenses)
});

export default connect(mapToStateStore)(ExpenseSammary);
