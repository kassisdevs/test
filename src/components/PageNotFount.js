import React from 'react';
import { Link } from 'react-router-dom';

const PageNotFount = () => (
  <div>404! <Link to="/" >Home</Link></div>
);

export default PageNotFount;