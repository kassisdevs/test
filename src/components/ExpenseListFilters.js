import React from 'react';
import { connect } from 'react-redux';
import { setTextInput } from '../actions/filters';
import { sortbyAmount, sortbyDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }
  onDatesChange = ({ startDate, endDate }) => {
  this.props.setStartDate(startDate)
  this.props.setEndDate(endDate)
  }
  onFocusChange = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }))
  }
  onTextChange = (e) => {
    this.props.setTextInput(e.target.value)
  }
  onFilterChange = (e) => {
    if (e.target.value === 'amount') {
     this.props.sortbyAmount()
    } else if (e.target.value === 'date') {
     this.props.sortbyDate()
    }
   }
  render(props) {
    return (
      <div>
      <input 
      type="text" 
      value={this.props.filters.text} 
      onChange={this.onTextChange} />
      <select 
      value={this.props.filters.sortBy} 
      onChange={this.onFilterChange}>
      <option value="date" >Date</option>
      <option value="amount" >Amount</option>
      </select>
      <DateRangePicker
      startDate={this.props.filters.startDate}
      endDate={this.props.filters.endDate}
      onDatesChange={this.onDatesChange}
      focusedInput={this.state.calenderFocused}
      onFocusChange={this.onFocusChange}
      showClearDates={true}
      numberOfMonths={1}
      isOutsideRange={() => false}
     />
      </div>
     );
  }
}


const mapStateToProps = (state) => ({
  filters: state.filters
  });

const mapDispatchToProps = (dispatch) => ({
  setTextInput: (text) => dispatch(setTextInput(text)),
  sortbyAmount: () => dispatch(sortbyAmount()),
  sortbyDate: () => dispatch(sortbyDate()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))


});

const ConnectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;