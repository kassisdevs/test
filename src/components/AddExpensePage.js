import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { ADD_EXPENSE } from '../actions/expenses';
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
      <ExpenseForm
      onSubmit={this.onSubmit}
      />
      </div>
     );
  }
};


 const mapTodispatch = (dispatch) => ({
  addExpense: (expense) => dispatch(ADD_EXPENSE(expense))
 });

 export default connect(undefined, mapTodispatch)(AddExpensePage);