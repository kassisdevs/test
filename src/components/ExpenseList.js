import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
 <div>
 {
   props.expenses.length === 0 ? (
      <p>no expenses</p>
   ) : (
      props.expenses.map(expense => (
      <ExpenseListItem key={expense.id} { ...expense } />
      ))
   )
 }
 
 </div>
);

const mapToStateStore = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
  };

export default connect(mapToStateStore)(ExpenseList);