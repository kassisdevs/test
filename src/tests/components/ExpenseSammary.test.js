import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSammary } from '../../components/ExpenseSummary';
import { expenses } from './../fixetures/expenses';
import selectTotalExpenses from '../../selectors/expenses-total';

test('Should return 0 expenses and 0 total', () => {
  const wrapper = shallow(<ExpenseSammary />);
  expect(wrapper).toMatchSnapshot();
});

test('Should return 3 expense and 3 total', () => {
  const expenseCount = expenses.length;
  const expensesTotal = selectTotalExpenses(expenses);
  const wrapper = shallow(<ExpenseSammary
    expenseCount={expenseCount}
    expensesTotal={expensesTotal}
  />);
  expect(wrapper).toMatchSnapshot();
});
