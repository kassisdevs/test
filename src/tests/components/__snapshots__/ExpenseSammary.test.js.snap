// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return 0 expenses and 0 total 1`] = `
<div>
  <p>
    Viewing 
     expenses totaling 
    $0.00
  </p>
</div>
`;

exports[`Should return 3 expense and 3 total 1`] = `
<div>
  <p>
    Viewing 
    3
     expenses totaling 
    $95.00
  </p>
</div>
`;
